Spinachboul update function (#457)
SciML/Surrogates.jl: docs/src/tensor_prod.md

# Tensor product function

> The tensor product function is defined as: ```\[ f(x) = ∏ᵢ=₁ᵈ cos(aπxᵢ) \]```

Reviewer: But this isn't correct. This is one example of a tensor product.

Reviewer 2: Yes, that is true. Maybe we specifically say it is one of the possible forms. I think this was taken from https://smt.readthedocs.io/en/latest/_src_docs/problems/tensorproduct.html

Author: Well, @ChrisRackauckas  and @sathvikbhagavan 
Should properly define the tensor product function just like this:
![equation](https://github.com/SciML/Surrogates.jl/assets/105979087/18d4958f-e343-4c72-a180-3556b131458d)

Or just stating this example would be fine for the documentation??

Reviewer: Yes that would do it. But do we have an actual issue that is being solved here?

Reviewer 2: As the benchmarks need some cleanup, I think we can startoff with this PR to clean up the tensor product benchmark? Although I am not sure why would we need a surrogate for computing tensor product :sweat_smile:

Author: @sathvikbhagavan 
I think there would be a couple of factors as to why surrogates could be used to compute tensor product:
- Computational Complexity would be decreased
- Efficient memory handling
- Normalization of products of tensors which provide numerical stability

This PR #457 only consists of changing the example of tensor product into a more readable form up till now, but I would also work towards explaining the mathematical concepts behind surrogate models in my draft PR as we discussed earlier.

# Tensor product function

> A tensor product function combines multiple functions or vectors using the tensor product operation. The tensor product is a mathematical operation that takes two vectors and produces another vector space, capturing their joint behavior across multiple dimensions.

Reviewer 2: ```suggestion
A tensor product combines multiple functions or vectors. It is a mathematical operation that takes two vector spaces and produces another vector space, capturing their joint behavior across multiple dimensions.
```

For instance, consider a tensor product function defined as follows:

> ```\[ f(x) = ∏ᵢ=₁ᵈ cos(aπxᵢ) \]```

Reviewer 2: Is this different from what it was before?

Author: No, it is the same just in a more readable form than before, the limits and the function, which we usually write!
As earlier it was a bit confusing as to what the actual function is!

Reviewer 2: Alright, then its fine

plot!(xs, loba_1.(xs), label="Lobachevsky", legend=:top)
plot!(xs, krig.(xs), label="Kriging", legend=:top)
```

## Kriging Plot

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: I didn't do anything new logically in the code, just added in some explanation about the tensor product function which I thought would be relevant and added the plots , i will remove if it is not required!!

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Reviewer 2: Yes, remove them

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: @sathvikbhagavan
But then, shouldn't we add more mathematical explanation to improve our readme file, simply stating the function would be too vague for the viewers!

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Reviewer 2: I don't understand the usefulness of the plots. They don't provide any new information either. Look at the current plot in the tutorial - https://docs.sciml.ai/Surrogates/stable/tensor_prod/

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Reviewer 2: Mathematical information can be improved in the description of each surrogate model. Benchmarks don't require explaining the models again.

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: @sathvikbhagavan 
In every, individual model, there is just the code written, like in this file: https://github.com/SciML/Surrogates.jl/blob/master/src/GEK.jl

So I believe we could mention the maths concept behind, let the users break-down each segment of the code.
In this case, the maths concept behind would be: **Gaussian Process Regression with Exponential Kernel**
Likewise we could explain about the topics **Gaussian Process** and **Exponential Kernel**

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Reviewer 2: Thats the source code. Users are not concerned with the internal details and they won't see the source code.
There are two things that are important - user facing API (things which are exported from the package) and user facing documentation.

I agree that mathematical concepts need to be explained in detail for various models which will be useful. Those things will go in pages like https://docs.sciml.ai/Surrogates/stable/radials/ which are tutorials explaining the model and an example on how to use it and their limitations etc. 

Benchmark problems are more concerned with for a problem how each model fares and what is the best among them.

You can start with cleaning up each model tutorial and add mathematical explanations etc.

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: Ok, so for now, I just need to focus on adding mathematical explanations in the tutorials right?
So soon I'll share a sample file (won't hit a PR for this) and then request your approval if it is going along the correct lines. 
Please then suggest to me the correct blend of theory and math explanations which would enable me to work efficiently.

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Reviewer 2: Sure. Feel free to open up a PR and continue working. Make it a draft PR and you can push in commits. When you are ready, make it ready for review and ping me. I will take a look. Hopefully you are more comfortable in the ecosystem and how things work in SciML.

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: Yeah! I am actually getting to know SciML codebase better. Thanks for your guidance throughout this time.

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: Hey @ChrisRackauckas  and @sathvikbhagavan 
Please let me know if this explanation suits well with the Kriging Model!

![pic2](https://github.com/SciML/Surrogates.jl/assets/105979087/4929aae0-70b8-4c5a-a933-a010fdb6a5bc)

![pic1](https://github.com/SciML/Surrogates.jl/assets/105979087/0c2a03e2-d698-4b1b-9194-4d5de454976e)

> ![kriging](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/906e6688-db47-48be-90d1-ea471aacac16)

Author: And please also suggest models on which I could work! I mean there are many out there but the ones which are important and really catch reader's eye.

## Lobachevsky Plot

![lobachevsky](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/678cfc13-0aec-4488-8e4d-39649853ecdd)

## Combined Plot

> ![combined_plot](https://github.com/Spinachboul/Surrogates.jl/assets/105979087/46762f0d-50c5-4d6c-961a-236fd9fb3ad5)

Reviewer: This is unrelated?

Reviewer: Oh I see, but the docs already generate these plots.

Author: Yeah, that's right! I attached the plots for reference about Kriging and Lobachevsky individually, and their plotting curves. And can't we attach plots both, in the readme and the docs??

Reviewer 2: No, I don't recommend attaching plots but instead running code to generate them
